File: scripts/merge-analyses.py
#!/usr/bin/env python

# This script merges the analysis data in the files provided as
# arguments, and prints the merged analysis data to stdout.
# The "target" data lines from the input files are copied to
# the output as-is. The "source" data lines are merged such
# that the syntax and symbol properties are unioned across
# all input lines that have a matching (loc, pretty) tuple.
# This ensures that for a given identifier, only a single context
# menu item will be displayed for a given "pretty" representation,
# and that context menu will link to all the symbols from all
# the input files that match that.

# Note that this code intentionally preserves the order of stuff
# within each line, as much as possible. However, the lines
# themselves get reordered, but the output can be piped through
# `sort` as needed. Maintaining the order of stuff within the
# line is important as reordering things can result in identical
# lines getting sorted differently. e.g.
#   { loc: foo, pretty: bar }
# and
#   { pretty: bar, loc: foo }
# are semantically identical but would not get deduplicated by
# piping through `sort` and `uniq`.

from collections import OrderedDict
import json
import sys

if len(sys.argv) == 1:
    sys.stderr.write("Usage: merge-analyses.py <filename> [<filename> ...]\n")
    sys.stderr.write("  This script will merge the analysis data from the given files\n")
    sys.stderr.write("  and print it to stdout\n")
    sys.exit(1)

sourcemap = {}

def SplitToOrderedSet(comma_sep_str):
    return OrderedDict([(item, None) for item in comma_sep_str.split(',')])

for filename in sys.argv[1:]:
    with open(filename) as filehandle:
        for line in filehandle.readlines():
            # parse the JSON into an OrderedDict to maintain order of properties
            entry = json.loads(line, object_pairs_hook=OrderedDict)

            if "source" not in entry:
                print(line.strip())
                continue

            for prop in entry:
                if prop not in ["source", "loc", "pretty", "syntax", "sym", "no_crossref"]:
                    sys.stderr.write("WARNING: Unexpected property %s found in source line in analysis file %s\n" % (prop, filename))

            key = (entry["loc"], entry["pretty"])
            if key in sourcemap:
                # We already encountered a line with this (loc,pretty) tuple,
                # so merge the syntax and sym properties. Again use an
                # OrderedDict (with None values, so it is effectively an
                # ordered set) to avoid reordering the existing tokens.
                oldEntry = sourcemap[key]
                syntaxes = SplitToOrderedSet(oldEntry["syntax"])
                syntaxes.update(SplitToOrderedSet(entry["syntax"]))
                oldEntry["syntax"] = ','.join(i for i in syntaxes)
                symbols = SplitToOrderedSet(oldEntry["sym"])
                symbols.update(SplitToOrderedSet(entry["sym"]))
                oldEntry["sym"] = ','.join(i for i in symbols)
                sourcemap[key] = oldEntry
            else:
                # Haven't encountered this before, so add it to the map
                sourcemap[key] = entry

for key in sourcemap:
    print(json.dumps(sourcemap[key], separators=(',', ':')))

File: tools/src/bin/merge-analyses.rs
//! This tool merges the analysis data in the files provided as arguments,
//! and prints the merged analysis data to stdout. The "target" data lines
//! from the input files are emitted to the output, but normalized and
//! deduplicated. The "source" data lines are merged such that the `syntax`
//! and `sym` properties are unioned across all input lines that have a
//! matching (loc, pretty) tuple.
//! This ensures that for a given identifier, only a single context menu
//! item will be displayed for a given "pretty" representation, and that
//! context menu will link to all the symbols from all the input files that
//! match that.
//!
//! Note that as this code uses the analysis.rs code for parsing and printing,
//! the emitted output should always be in a consistent/normalized format.

use std::env;
use std::io::stdout;

extern crate regex;
use regex::Regex;

extern crate env_logger;

extern crate tools;
use tools::file_format::merger::merge_files;

fn main() {
    env_logger::init();

    let args: Vec<_> = env::args().skip(1).collect();

    if args.len() == 0 {
        eprintln!("Usage: merge-analyses <filename> [<filename> [...]]");
        eprintln!("  This tool will merge the analysis data from the given files");
        eprintln!("  and print it to stdout; each line will be in a normalized format.");
        std::process::exit(1);
    }

    // The paths are relative, so don't look for a leading slash, but instead anchor at the front.
    let re_platform = Regex::new(r"^analysis-([^/]+)/").unwrap();

    // Build a list of platforms that parallels the list of files in `args`.
    let platforms: Vec<String> = args
        .iter()
        .enumerate()
        .map(|(i, fname)| {
            re_platform
                .captures(fname)
                .and_then(|c| c.get(1))
                .map_or(format!("platform-{}", i), |m| m.as_str().to_string())
        })
        .collect();

    merge_files(&args, &platforms, &mut stdout());
}
